substrate/client/transaction-pool/src/fork_aware_txpool/view_store.rs

```diff
@@ -650,6 +654,71 @@ where
 		log::trace!(target:LOG_TARGET,"finish_background_revalidations took {:?}", start.elapsed());
 	}
 
+	/// Reports invalid transactions to the view store.
+	///
+	/// This function accepts an array of tuples, each containing a transaction hash and an
+	/// optional error encountered during the transaction execution at a specific (also optional)
+	/// block.
+	///
+	/// Removal operation applies to provided transactions and their descendants.
+	///
+	/// Invalid future and stale transaction will be removed only from given `at` view, and will be
+	/// kept in the transaction pool. Such transaction will not be reported in returned vector. They
+	/// also will not be banned. No event will be triggered.
+	///
+	/// For other errors, the transaction will be removed from the pool, and it will be included in
+	/// the returned vector. It will be included in the returned vector. Additionally transactions
+	/// provided as input will be banned from re-entering the pool.
+	///
+	/// If the tuple's error is None, the transaction will be forcibly removed from the pool,
+	/// banned and included into the returned vector.
+	///
+	/// For every transaction removed from the pool (including descendants) an Invalid event is
+	/// triggered.
+	///
+	/// Returns the list of actually removed transactions, which may include transactions dependent
+	/// on the provided set.
+	pub(crate) fn report_invalid(
+		&self,
+		at: Option<Block::Hash>,
+		invalid_tx_errors: &[(ExtrinsicHash<ChainApi>, Option<BlockchainError>)],
+	) -> Vec<TransactionFor<ChainApi>> {
+		let mut remove_from_view = vec![];
+		let mut remove_from_pool = vec![];
+
+		invalid_tx_errors.iter().for_each(|(hash, e)| match e {
+			Some(BlockchainError::ApplyExtrinsicFailed(ApplyExtrinsicFailed::Validity(
+				TransactionValidityError::Invalid(
+					InvalidTransaction::Future | InvalidTransaction::Stale,
+				),
+			))) => {
+				remove_from_view.push(*hash);
+			},
+			_ => {
+				remove_from_pool.push(*hash);
+			},
+		});
+
+		// transaction removed from view, won't be included into the final result, as they may still
+		// be in the pool.
+		at.map(|at| {
+			self.get_view_at(at, true)
+				.map(|(view, _)| view.remove_invalid(&remove_from_view[..]))
```

Comment:
The docs of `report_invalid` say that tx that are stale or future will not be banned and no event will be sent. However, validated pool will indeed ban the tx down the callchain chain.

Reply:
```suggestion
				.map(|(view, _)| view.remove_invalid(&remove_from_view))
```
We shouldn't need this.

Reply:
> The docs of `report_invalid` say that tx that are stale or future will not be banned and no event will be sent. However, validated pool will indeed ban the tx down the callchain chain.

Good catch.

The tx will be banned only on this particular view, which makes some sense - we probably don't want this transaction on this fork.

For transaction being removed with `remove_transaction_subtree` it will be banned for every view.

For now it should work good enough. To improve this maybe we should consider introducing a rotator at the pool level?

Reply:
Clarified in 4200ccc.